tmt/base.py

```diff
@@ -1435,6 +1435,22 @@ def name_filter(nodes):
                 node for node in nodes
                 if any([re.search(name, node.name) for name in cmd_line_names])]
 
+        # Get tests whose property 'enabled' is true (when option --enabled is specified)
+        if Test._opt('enabled'):
```

Comment:
I think this doesn't work as intended. Filters is a  list however each item may contain more values joined by `&` or `|`.

So let's say I execute `tmt t ls --filter 'test:echo&enabled:true'` then `filters=['test:echo&enabled:true']` is used. So for `tmt t ls --filter 'test:echo&enabled:true' --disabled` it becomes `filters=['test:echo&enabled:true', 'enabled:false']` which obviously prints nothing as my cmdline 'enabled:true' was not removed.

If the preferred way to achieve disable/enable is using filter then IMO tmt needs to remove any `enabled:XXX` regardless the value (there is '-' for not so "enabled:-false" is valid and working) and remove & or | used to for "enabled" as well. IIRC we allow anything in the value, but please confirm in fmf.filter implementation, so value is either until the end of the string or & or |. This removal could cause empty item in filters so that should be addressed too. 

So for my `filters=['test:echo&enabled:true']` it would remove enabled to get `filters=['test:echo]` (& has to be removed as well) and later add new item so `filters=['test:echo', 'enabled:false']` is the final filter.

We don't have priorities for AND and OR so just remove that character before 'enabled:' from the filter string and it should be OK.


Also would be best to have single place for this filter transformation method and use it from all places (easier to test via unit test etc.)

Reply:
To handle both `&`  and `|`, looks it is a bit complicated to remove `&, |` from `filters=['test:echo&enabled:true']` or `filters=['test:echo|enabled:true']`. Currently `&` and `|` is not handled in `Tree.plans()`, `Tree.tests()` and `Tree.stories()`. e.g.
```python
$ python3 -m pdb ~/.virtualenvs/tmt/bin/tmt \
   tests show --filter 'enabled:true|test:echo'
(Pdb) b tmt/base.py:1439
Breakpoint 1 at /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/base.py:1439
(Pdb) c
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/base.py(1439)tests()
-> if Test._opt('enabled'):
(Pdb) l
1434 	            return [
1435 	                node for node in nodes
1436 	                if any([re.search(name, node.name) for name in cmd_line_names])]
1437 	
1438 	        # Get tests whose property 'enabled' is true (when option --enabled is specified)
1439B->	        if Test._opt('enabled'):
1440 	            # '--enabled' win '--filter enabled:false'
1441 	            while 'enabled:false' in filters:
1442 	                filters.remove('enabled:false')
1443 	            if 'enabled:true' not in filters:
1444 	                filters.append('enabled:true')
(Pdb) p filters
['enabled:true|test:echo']
(Pdb) c
/tests/disabled2
     summary Test tmt tests ls/show --disabled
        test echo
        path /tests
      manual false
    duration 5m
     enabled false
      result respect

/tests/enabled1
     summary Test tmt tests ls/show --enabled
        test /bin/true
        path /tests
      manual false
    duration 5m
     enabled true
      result respect

/tests/enabled2
     summary Test tmt tests ls/show --enabled
        test echo
        path /tests
      manual false
    duration 5m
     enabled true
      result respect

The program exited via sys.exit(). Exit status: 0
```

Hence, we can use `re.sub()` to replace those filters including `enabled:(true|false)`. e.g.
```python
$ cat -n /tmp/foo.py 
     1	#!/usr/bin/python3
     2	import sys
     3	import re
     4	
     5	
     6	def revise_filters(filters):
     7	    out = []
     8	    for filter in filters:
     9	        if 'enabled:' in filter:
    10	            filter = re.sub(r'enabled:(true|false)', 'enabled:true', filter)
    11	        out.append(filter)
    12	    return out
    13	
    14	
    15	def main(argc, argv):
    16	    print(revise_filters(argv[1:]))
    17	    return 0
    18	
    19	
    20	if __name__ == '__main__':
    21	    sys.exit(main(len(sys.argv), sys.argv))
$ 
$ /tmp/foo.py 'test:echo&enabled:false'
['test:echo&enabled:true']
$ /tmp/foo.py 'enabled:false|test:echo'
['enabled:true|test:echo']
``` 

@lukaszachy, what do you think?

Reply:
Agreed that it's not necessary to handle the conflicting `--filter` content with `--enabled` and `--disabled`. Let's just append the additional filter.